scalalib/src/mill/javalib/android/AndroidAppModule.scala

```diff
@@ -0,0 +1,333 @@
+          androidSdkModule.androidJarPath().path.toString // Include Android framework classes
+        )
+    )
+
+    PathRef(dexOutputDir)
+  }
+
+  /**
+   * Packages the DEX files and Android resources into an unsigned APK using the `aapt` tool.
+   *
+   * The `aapt` tool takes the DEX files (compiled code) and resources (such as layouts and assets),
+   * and packages them into an APK (Android Package) file. This APK file is unsigned and requires
+   * further processing to be distributed.
+   *
+   * @return A `PathRef` pointing to the unsigned APK file.
+   */
+  def androidApk: T[PathRef] = Task {
+    val unsignedApk: os.Path = T.dest / s"${androidAppName().toString}.unsigned.apk"
+
+    os.call(
+      Seq(
+        androidSdkModule.aaptPath().path.toString,
+        "package",
+        "-f",
+        "-M",
+        (millSourcePath / "src/main/AndroidManifest.xml").toString, // Path to AndroidManifest.xml
+        "-I",
+        androidSdkModule.androidJarPath().path.toString, // Include Android JAR
+        "-F",
+        unsignedApk.toString // Output APK
+      ) ++ Seq(androidDex().path.toString) // Include DEX files
+    )
+
+    PathRef(unsignedApk)
+  }
+
+  /**
+   * Optimizes the APK using the `zipalign` tool for better performance.
+   *
+   * The `zipalign` tool ensures that all uncompressed data in the APK is aligned on a 4-byte boundary.
+   * This is required for better performance on Android devices. This step is done after the APK
+   * is created but before it is signed.
+   *
+   * For more details on the zipalign tool, refer to:
+   * [[https://developer.android.com/tools/zipalign zipalign Documentation]]
+   *
+   * @return A `PathRef` pointing to the aligned APK file.
+   */
+  def androidAlignApk: T[PathRef] = Task {
+    val alignedApk: os.Path = T.dest / s"${androidAppName().toString}.aligned.apk"
+
+    os.call(
+      Seq(
+        androidSdkModule.zipalignPath().path.toString, // Call zipalign tool
+        "-f",
+        "-p",
+        "4", // Force overwrite, align with 4-byte boundary
+        androidApk().path.toString, // Use the unsigned APK
+        alignedApk.toString // Output aligned APK
+      )
+    )
+
+    PathRef(alignedApk)
+  }
+
+  /**
+   * Signs the APK using a keystore to generate a final, distributable APK.
+   *
+   * The signing step is mandatory to distribute Android applications. It adds a cryptographic
+   * signature to the APK, verifying its authenticity. This method uses the `apksigner` tool
+   * along with a keystore file to sign the APK.
+   *
+   * If no keystore is available, a new one is generated using the `keytool` utility.
+   *
+   * For more details on the apksigner tool, refer to:
+   * [[https://developer.android.com/tools/apksigner apksigner Documentation]]
+   *
+   * @return A `PathRef` pointing to the signed APK.
+   */
+  def androidApp: T[PathRef] = Task {
+    val signedApk: os.Path = millSourcePath / s"${androidAppName().toString}.apk"
```

Comment:
We should not be writing to `millSourcePath`. This should live in `T.dest` like any other generated file; if the user isn't sure where it is, they can always use `mill show`

Reply:
done